keps/sig-network/1880-services-ip-ranges/README.md

```diff
@@ -0,0 +1,1015 @@
+Service is not created, the IPaddress will be orphan. To avoid leaks, a controller will use the
+`metadata.creationTimestamp` field to detect orphan objects and delete them.
+
+There is going to be a controller to avoid leaking resources:
+
+- checking that the corresponding parentReference on the IPAddress match the corresponding Service
+- cleaning all IPAddress without an owner reference or, if the time since it was created is greater
+than 60 seconds (default timeout value on the kube-apiserver )
+
+
+#### Edge cases
+
+Since we have to maintain 3 relationships Services, ServiceIPRanges and IPAddresses, we should be able
+to handle edge cases and race conditions.
+
+- Valid Service and IPAddress without ServiceIPRange:
+
+This situation can happen if a user forcefully delete the ServiceIPRange, it will be recreated for the
+"default" ServiceIPRanges because the information is in the apiserver flags, but for other ServiceIPRanges
+that information is no longer available.
+
+Another possible situation is when one ServiceIPRange has been deleted, but the information takes too long
+to reach one of the apiservers, its allocator will still consider the range valid and may allocate one IP
+from there. To mitigate this problem, we'll set a grace period of 60 seconds on the ServiceIPrange controller
+to remove the finalizer, if an IP address is created during this time we'll be able to block the deletion
+and inform the user.
+
+[![](https://mermaid.ink/img/pako:eNptkTFvwyAQhf8KYm2SxlZVtQyRbNMhW9WsLCe4JEg2uBhaVVH-eyG4dZyWASHed--e7k5UWoWU0QHfAxqJXMPBQScMiQekt47UBAbCsUWPZIfuQ0tstvwtIz04r6XuwXjSJBB6PUQIHan-EnxO1NdtqqQ1DmFqI0wGarLcbO4a9k8KUjyXq-LxabVere_Lh8znu0pVhLMb01E1Nn7ZlCIS0LZWJmZyK8rM8YtLxX7Ls5vKapPUZc1mkdQl5USMMeZJ59SU5qrT9pV8an-0wd_MXRi6oB26DrSKuzslD0H9ETsUlMWnwj2E1gsqzDmioVcx8ovScc6U7aEdcEEheLv7MpIy7wL-QOP-R-r8DRVXqSY)](https://mermaid.live/edit#pako:eNptkTFvwyAQhf8KYm2SxlZVtQyRbNMhW9WsLCe4JEg2uBhaVVH-eyG4dZyWASHed--e7k5UWoWU0QHfAxqJXMPBQScMiQekt47UBAbCsUWPZIfuQ0tstvwtIz04r6XuwXjSJBB6PUQIHan-EnxO1NdtqqQ1DmFqI0wGarLcbO4a9k8KUjyXq-LxabVere_Lh8znu0pVhLMb01E1Nn7ZlCIS0LZWJmZyK8rM8YtLxX7Ls5vKapPUZc1mkdQl5USMMeZJ59SU5qrT9pV8an-0wd_MXRi6oB26DrSKuzslD0H9ETsUlMWnwj2E1gsqzDmioVcx8ovScc6U7aEdcEEheLv7MpIy7wL-QOP-R-r8DRVXqSY)
+
+For any Service and IPAddress that doesn't belong to a ServiceIPRange the controller will raise an event
+informing the user, keeping the same behavior
+
+```go
+// cluster IP is out of range
+c.recorder.Eventf(&svc, nil, v1.EventTypeWarning, "ClusterIPOutOfRange", "ClusterIPAllocation", "Cluster IP [%v]:%s is not within the service CIDR %s; please recreate service",
+```
+
+- Valid Service and ServiceIPRange but not IPAddress
+
+It can happen that an user forcefully delete an IPAddress, in this case, the controller will regenerate the IPAddress.
+
+During this time, there is a chance that an apiserver tries to allocate this IPAddress, with a possible situation where
```

Comment:
This is a problematic answer (and I know I helped make it :)

Keeping it in memory on the apiserver only works as long as the apiserver doesn't restart.  IOW, we can't *really* guarantee it, unless we keep YET ANOTHER object like the hidden bitmap.  I wrote up a lot and then realized this, so I will leave my comment here, but I don't think it is valid.

For Services we can fix it, but for opaque other thingds we can't.

Perhaps we should not try to "really" fix it.  You went behind our back and force-deleted an IP or your controller did not release the IP properly.  You get what you paid for.


------------------Probably garbage below----------------------

This is a place where UID might matter.  Consider this:

Create a Gateway (or something you don't really know about) that allocates an IP.  They set the ref to their own kind and name.  Someone deletes that IP.  apiserver sees the DELETE but the thing referenced still exists, so it leaves it marked as "terminating".  We can periodically re-check, but we are stuck like this forever.

What REALLY happened was:
* User created Gateway "foo"
* Gateway controller allocated IP and ref -> "foo"
* User deleted gateway "foo"
* gateway controller deletes the IP
* Race:
  - user creates a new Gateway "foo"
  - apiserver observes IP deletion

Or even worse:
* User created Gateway "foo"
* Gateway controller allocated IP and ref -> "foo"
* User deleted gateway "foo"
* gateway controller doesn't delete the IP (leave it for GC)
* user creates a new Gateway "foo"
* apiserver repair loop finds the IP with a valid ref to "foo"
  
If the new gateway is created before the apiserver observes the delete, apiserver will find that gateway "foo" still exists and can't release the IP.  It can't peek inside "foo" to see if that is the right IP because it is a type it does not know.  If it knew the UID it could see that "foo" UID was different and release the IP.

I see that we do not have a UID at BeginCreate.  That seems wrong to me.  Will ask api-machinery.

Reply:
> user creates a new Gateway "foo"
> apiserver repair loop finds the IP with a valid ref to "foo"

that newly created Gateway will have a new IP, the repair loop will find 2 IPs to the same Object and we are in the situation you've described here https://github.com/kubernetes/enhancements/pull/3365/files#r900423953

I think we've agreed to move the repair from the apiservers to a single controller on the controller manager.

Reply:
UID fixes this :)

> I think we've agreed to move the repair from the apiservers to a single controller on the controller manager.

Maybe we should do that as a distinct step?